flang/lib/Lower/Bridge.cpp

```diff
@@ -1240,6 +1262,43 @@ class FirConverter : public Fortran::lower::AbstractConverter {
                                        valueList, blockList);
   }
 
+  void genFIR(const Fortran::parser::AssociateConstruct &) {
+    for (auto &e : getEval().getNestedEvaluations())
+      genFIR(e);
+  }
+
+  void genFIR(const Fortran::parser::AssociateStmt &stmt) {
+    Fortran::lower::StatementContext stmtCtx;
+    localSymbols.pushScope();
+    for (auto &assoc :
+         std::get<std::list<Fortran::parser::Association>>(stmt.t)) {
+      auto &sym = *std::get<Fortran::parser::Name>(assoc.t).symbol;
+      const auto &rhsExpr =
+          *sym.get<Fortran::semantics::AssocEntityDetails>().expr();
+      if (!Fortran::evaluate::IsVariable(rhsExpr) ||
+          Fortran::evaluate::HasVectorSubscript(rhsExpr)) {
+        // Associating entity is assigned the value of the selector expression.
+        mapSymbolAttributes(Fortran::lower::pft::Variable(sym), stmtCtx);
```

Comment:
Is the new code in line with what you had in mind?

Reply:
It is more in lined with what I had in mind, though I think there is an even easier path. See my reply in Eric's comment.
With your approach, the `createTemporary` properly if the expression has non compile time constant length/extents.
Then `addSymbol` is only OK with non character intrinsic scalar.
Also, I think `createTemporary` creates stack temps, for array, we would most likely want heap allocation, which implies some deallocation on exit of the associate construct.
Solving these problems will bring other troubles that I think you could avoid by using `genExprAddr` directly, instead of managing the temp creation/assignment here.